token-based greasing / initial packet protection

kazuho opened the pull request:
Based on the discussion in #3111, this PR implements greasing using the NEW_TOKEN frame as the conveyor.

It defines defines three properties: Alternative Version Number, Packet Type Modifier, Initial Salt, for obfuscating the version number field and the long header packet type, as well as encrypting the Initial packet payload.

Those three properties are carried by a NEW_TOKEN frame, and a client uses those "alternative initial sets" when resuming a connection using a NEW_TOKEN token. A server that is not interested in greasing or encrypting the Initial packets can advertise the values defined in the specification as the "alternative initial sets," and the client behavior would be the same as it is now (with the only difference being that the format of NEW_TOKEN frame has changed).

The text discusses separately about the properties and how they are distributed. Therefore it would be straightforward to define other methods of distributing the "alternative initial set" (e.g., by using the alt-svc header field).

Closes #2496, closes #3111, an alternative to #2573.

kazuho commented:
@marten-seemann I'm not sure if that actually works.

Because how a token is embedded in an Initial packet depends on the actual version of the QUIC packet that the server is processing. Therefore, I think that the server needs to first consult the version number on the wire, then use that to determine where the token is to decrypt the initial salt contained in the token.

Or are you suggesting something different?

DavidSchinazi commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2721,66 @@ between endpoints.  Application protocols SHOULD define rules for handling
+  MUST NOT be a reserved version ({{versions}}).
+
+* Packet type modifier; a two-bit value that is to be applied as a bit-wise
+  exclusive or (XOR) to the most significant bits of the Initial, Handshake,
```
Shouldn't this be `to the Long Packet Type field of` instead of `to the most significant bits of` ?

kazuho replied:
Definitely Yes! I missed the fact that we had a name for the two bits. Thank you.

martinthomson replied:
We have two choices for the placement of these bits in the NEW_TOKEN.  0bXXTTXXXX or 0bXXXXXXTT.  The former might be easier to apply, the latter might be easier to explain.

kazuho replied:
Fixed in 7eded0a.

DavidSchinazi commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2721,66 @@ between endpoints.  Application protocols SHOULD define rules for handling
+packet type modifier chosen at random.  Then, when issuing a token using a
+NEW_TOKEN frame, the server generates the alternative initial salt by calling a
+pseudo-random function, embeds that initial salt into the token which is then
+encrypted, and sends a NEW_TOKEN frame that comprises of the generated token and
```
I might be worth adding a bit more detail to how the token is encrypted, perhaps `which is then encrypted using a key only known to the server`

kazuho replied:
Thanks. Makes sense. Adopted in d16700d.

DavidSchinazi commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2721,66 @@ between endpoints.  Application protocols SHOULD define rules for handling
+decrypt the payload of the Initial packet.
+
+When the server is incapable of determining the alternative initial salt, it can
+send a Version Negotiation packet that instructs the client to use the default
```
In the current spec, the client fails the connection upon receipt of a Version Negotiation packet. So this has a dependency on the version downgrade prevention extension.

But assuming the presence of a version negotiation mechanism with downgrade prevention, I think we have an issue here. If the server rotates the key it uses to encrypt the alternative initial salt, and then receives a token it can no longer decrypt, it sends a version negotiation packet, and the client tries again with QUICv1. This is likely to trip the version downgrade prevention mechanism - how do we distinguish between a server that has forgotten their key and an attacker that is trying to force a downgrade to QUICv1?

kazuho replied:
This PR has a built-in downgrade prevention mechanism.

The requirement we have in this PR is that each NEW_TOKEN token that the server issues is linked to a particular alternative version number.

When a server receives an Initial packet with an alternative version number it recognizes but with a token it cannot use, the server sends a VN. The client reconnects using V1, but the text in the PR requires the client to continue using the token. Then, when the server receives the V1 packet with the token, and can in fact use the token and finds out that the token was in fact being issued for a version other than V1, the server generates a VERSION_NEGOTIATION_ERROR.

In short, it is a downgrade attack if VN was issued while the token was usable.

DavidSchinazi replied:
Let's assume the presence of an on-path attacker that can drop, forge and rewrite packets. This attacker is interested in parsing the SNI and ALPN. The attacker lets the client's first QUICv1 connection through, and the client gets a token with alternate version and salts, under the cover of encryption. Some time later, the client wants to open a second connection and sends an obfuscated initial with the new version and salt. The attacker can pretty easily figure out that the first packet is an Initial even with the 2bit XOR. The attacker modifies some bits in the token, and recomputes the checksum then sends the modified packet to the server, leaving the version intact. The server verifies the checksum then fails to decrypt the token. The server sends VN. The client then retries with QUICv1 using its original token. The attacker performs the same modification on that token, and recomputes the checksum and initial auth tag, then sends the packet to the server. The server will again fail to decrypt the token, it sends a retry, the client tries again with the retry token, which the attacker leaves alone, and the connection successfully establishes with QUICv1. The **attacker has successfully forced a downgrade** to QUICv1 and was able to extract SNI and ALPN.

martinthomson replied:
Yes, this offers an almost trivial downgrade attack, but this isn't a scenario that we agreed to cover.  It does seem to be a little bit *too* trivial an attack though.

kazuho replied:
@DavidSchinazi That's obviously the case. Thank you for pointing that out.

I think what the PR fails to state that a Retry token needs to inherit the properties of the NEW_TOKEN token included in the Initial packet that trigerred the Retry. One way doing that is to embed the NEW_TOKEN token in the Retry token.

The reason why we need the NEW_TOKEN token to be embedded in the Retry token is not only because there is an attack that you describe. It is necessary because when the server receives the first Initial packet from client that carries the alternative version, and responds with a Retry packet that carries that alternative version, and the client in response sends an Initial packet that carries the Retry token using an Initial packet with the alternative version, the server needs to extract the alternative initial salt (and other information) from the Retry token.

I'd update the PR to clarify that.

kazuho replied:
@martinthomson I think the design principle behind this PR that I have are:
* passive attack does not reveal the payload
* active attack can lead to the payload being revealed, but is detectable by the endpoints and thereby disrupts the handshake
* handshake falls back to v1 when the server forgets the token encryption key

As I think the issue pointed out by @DavidSchinazi is addressable (see my comment above), I think we might stick to the principle and how it evolves. If we conclude that the design based on the principle is simple enough, then we'd be fine. If not, we can consider the alternatives.

kazuho replied:
373257b clarifies that the alternative initial set would (or should) be available through the Retry token. The text is kept informative, as the PR no longer allows a downgrade.

DavidSchinazi commented on draft-ietf-quic-transport.md:
```diff
@@ -3940,6 +4007,19 @@ described in {{QUIC-TLS}}.  This protection does not provide confidentiality or
 integrity against on-path attackers, but provides some level of protection
 against off-path attackers.
 
+Additionally, the token is accompanied by a checksum.  This is because when a
```
Seeing a checksum here makes me uncomfortable. We're adding a checksum to retry packets because they have no other mechanism to detect corruption. Initial packets on the other hand have an auth tag, so a corrupted token will cause the deprotection to fail. In an ideal world we should handle that like any other corruption on the initial packet. But I think this goes back to my other question about version negotiation, I think there be dragons.

MikeBishop replied:
I second the question here; the checksum doesn't appear to serve a purpose not already covered here.

kazuho replied:
We need Token Checksum so that the server can easily determine if there was a bit flip on the wire when it receives an invalid token.

The PR (up until of 7eded0a) requires the server to drop the packet if there was a bit flip on the wire. OTOH, if it was an unusable token (rather than a bit flip), the server responds with a VN. This distinction is necessary to prevent bit flips from being mis-detected as an downgrade attack (see https://github.com/quicwg/base-drafts/pull/3166#discussion_r340344021).

That said, I agree that this change is intrusive. Considering that we need to require certain properties for the token (see https://github.com/quicwg/base-drafts/pull/3166#discussion_r340396154), we might as well state that a NEW_TOKEN token should be generated in a way that the server can detect bit flips even it has lost the keys to decrypt the token. In practice, that can be easily achieved by attaching CRC32 (or zero-key GMAC).

But finally, this entire discussion is based on the assumption that we need a downgrade path, which @martinthomson disputes in https://github.com/quicwg/base-drafts/pull/3166#pullrequestreview-308948794.

mikkelfj replied:
The checksum is very useful in situations where a bridge forwards a connection over an alternative medium such as USB or Bluetooth. The Retry and VN packets are the only packets that don't have self-integrity checks and such a bridge would be forced to provide additional framing. This framing is tolerable for VN because it is universal, but for Retry it requires version knowledge. It is much simpler if packets can just be forwarded with no knowledge other than length.

martinthomson replied:
The checksum can be provided by the server on its own.  The only reason you would provide a separate field for one is so that the client can validate it.  But the client can ensure that the NEW_TOKEN was not modified through packet protection.  Good to see this was removed :)

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2721,66 @@ between endpoints.  Application protocols SHOULD define rules for handling
+* Packet type modifier; a two-bit value that is to be applied as a bit-wise
+  exclusive or (XOR) to the most significant bits of the Initial, Handshake,
+  0-RTT, and Retry packets. This XOR is applied after the packets are encrypted
+  and before they are decrypted.
```
So... This isn't totally clear.  There are two "encryption" or "decryption" operations that might apply: the AEAD for packet protection and the masking for header protection.

I think that you mean to say that the value you feed into the AAD for packet protection is unmodified and this alteration only applies afterwards.

However, I would prefer it if were applied to the raw values so that the XOR is also authenticated by packet protection.  Then you have a clear order of operations for applying this information that matches the order of fields:

0. encode the packet
1. replace the version
2. mask the type (if long header)
3. apply packet and header protection using the new salt

kazuho replied:
Thanks for arguing for that. I had somehow thought that that's impossible, but I do not know why. Addressed in 14469be.

kazuho commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2721,66 @@ between endpoints.  Application protocols SHOULD define rules for handling
+packet type modifier chosen at random.  Then, when issuing a token using a
+NEW_TOKEN frame, the server generates the alternative initial salt by calling a
+pseudo-random function, embeds that initial salt into the token which is then
+encrypted, and sends a NEW_TOKEN frame that comprises of the generated token and
```
Applied in 6158bd9.

kazuho commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2721,66 @@ between endpoints.  Application protocols SHOULD define rules for handling
+
+A server advertises these values using a NEW_TOKEN frame {{frame-new-token}}.
+The token MUST include or associated with the alternative version number with
+which it can be used.
```
Thank you for the suggestions. Make sense. Applied in 06bd108.

kazuho commented:
@martinthomson 
> This is a whole lot easier if we force the server to commit, because then the server has to manage the fallback case. It makes this more brittle, in the sense that you need to make the static key more durable, but I think that it would be reasonable to expect a server to do that. They probably have to manage session ticket keys in this way.

I agree that it would be a much easier if we require the servers to commit.

At the same time, I think session ticket encryption keys are not something we should compare with. This is because the impact of losing access to alternative initial salt leads to handshake timeout, while the loss of session ticket encryption key merely leads to a full handshake. We have had this discussion in ESNI, and we ended up adding a fallback (i.e. esni_retry_request) due to a request from one of the largest deployments of TLS. That's why I have so far thought that a fallback is necessary.

That said, it might be possible that we have different requirements than ESNI.

After all, this alternative set thing is about giving the servers the *option* to issue different initial salts to every recurring connection. A server that advertises an alternative set does not need to work that way; deployments that might lose the key can issue a static initial salt that works across all the connections, and that would still be a win for us.

Considering these aspects, I am tempted to drop downgrade from this PR and see how it goes.

kazuho commented:
@DavidSchinazi @marten-seemann @martinthomson @MikeBishop @mikkelfj Thank you all for your feedback. I think that the PR is ready for a new round of review.

I've removed downgrade as well as the token checksum field (https://github.com/quicwg/base-drafts/pull/3166#discussion_r340432382 clarifies why it is no longer necessary), and added a Lifetime field to the NEW_TOKEN frame.

I'm still unsure if having no downgrade is a good idea, though I agree that this is much simpler to analyze.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -2740,8 +2740,10 @@ which is comprised of:
-The token MUST include or associated with the alternative version number with
-which it can be used.
+The token MUST permit the server to recover at least the alternative version
+number being associated to the token.  This property can be achieved for example
```
It MUST be able to recover the salt as well, I think.  It's possible that version number maps directly to salt -- but then, that's one implementation of being able to recover it.

kazuho replied:
Good point! Fixed in 757e5e0.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -4970,6 +5080,30 @@ The NEW_TOKEN frame is as follows:
+Lifetime:
+
+: Indicates the lifetime of the values contained in this frame in milliseconds.
+  An endpoint MUST NOT use the values provided by this frame, once the time that
```
```suggestion
  An endpoint MUST NOT use the values provided by this frame once the time that
```

mikkelfj commented:
I don't think we should weaken the core protocol just because H3 (usually) needs TCP fallback.

DavidSchinazi commented:
If I understand the latest text correctly, it's saying:
- NEW_TOKEN carries a lifetime
- if a server wants to use alternative salts, it MUST ensure that it has access to its decryption key for the lifetime of that token, otherwise handshake timeouts could occur

I don't think that'll work for our deployment. We're interested in deploying alternative salts, but our mechanism to share keys across our fleet of servers isn't infallible: it's possible that one server gets a new key before another server, and if a client hits these two servers in the wrong order, they could end up in a state where the server cannot decrypt the token. This is fine for the current uses of QUIC NEW_TOKENs and TLS session tickets, because both of those will resolve themselves in-band (via RETRY, or full handshake).

Therefore I really think we need a graceful mechanism to handle this, similar to what was added to ESNI. The question then becomes: can we design a fallback mechanism that isn't vulnerable to a simple downgrade attack.

Here's a straw-man:
- when the server sees an initial token (the one from NEW_TOKEN, not from retry) that it cannot decrypt, it sends a RETRY.
- the server takes the non-decryptable token it received from the client, and sends it encrypted inside the retry token in the server's RETRY 
- the client tries again, but this time with QUICv1 (no obfuscation) and the retry token, but we add a new transport parameter where the client sends the first token it used on this connection

This would allow the server to detect when someone messed with the token in flight, and would then close the connection. So it mitigates the downgrade attack I described [earlier](https://github.com/quicwg/base-drafts/pull/3166#discussion_r340366571).

This straw-man has many downsides, including:
- we're back to needing a checksum on the token
- an attacker can still see the SNI, but they can't do so without it being detected. This might be acceptable because the attacker can still force a downgrade to TLS/TCP today

Anyway, I'm not sure this straw-man is the way to go, but perhaps the idea of repeating tokens in transport parameters might be worth exploring to help detect this class of attack.

kazuho commented:
@DavidSchinazi Re downgrade prevention, I think we are converging on the same design (my design being the one that I had on this PR before dropping the downgrade).

Regarding the attack, I do not think we need to require the original token to be sent in a TP. Let me explain why:
* This PR requires the server that sends a Retry token to embed the the alternative sets of a NEW_TOKEN token in the Retry token (i.e. typically, NEW_TOKEN token found in an Initial packet that triggered the Retry would be embedded as-is in the Retry token).
* This PR used to require the client to continue sending the NEW_TOKEN token when downgrade happens. We can resurrect that requirement.

If we combine these two requirements, it would be possible for the server to detect a downgrade even if Retry was sent after downgrade, as the Retry token would convey the NEW_TOKEN token that the client originally submitted.

Re checksum, one way to solve this is to suggest servers advertising alternatives sets to embed a checksum to the tokens, using an key-less algorithm. For example, a server can build a NEW_TOKEN token as a combination of AES-CBC (encryption) and SHA256 (checksum).

Anyways, I think I might open a separate PR that describes the design with downgrade prevention, as having a PR would help us discuss the issues.

kazuho commented:
So I've created #3178 that adds downgradability on top of this PR, hoping that we can figure out a downgradable design that works and consider the costs.

mikkelfj commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2724,96 @@ between endpoints.  Application protocols SHOULD define rules for handling
+  modifier chosen at random.
+
+* When issuing a NEW_TOKEN token, the server generates the alternative initial
+  salt by calling a pseudo-random function.  Then it builds a token that
```
There is no need to use a pseudo-random function if the server has access to better randomness (but to be fair this is for a typical implementation). I would just say it generates a cryptographically secure random number.

kazuho replied:
Yes, the intent here is to recommend CSPRNG. Applied in febd899.

mikkelfj commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2724,96 @@ between endpoints.  Application protocols SHOULD define rules for handling
+* When issuing a NEW_TOKEN token, the server generates the alternative initial
+  salt by calling a pseudo-random function.  Then it builds a token that
+  embeds the alternative seeds including the initial salt being generated.  The
+  token will be encrypted using a key known only to the server, thereby
```
```suggestion
  token will be encrypted using a key known only to the server, or server cluster, thereby
```

kazuho replied:
Thank you for the suggestions. All the proposed editorial changes made sense except this one, I've applied them.

Regarding this suggestion, I think we do not need to be general here, as this is about how a "typical server" would behave.

mikkelfj commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2724,96 @@ between endpoints.  Application protocols SHOULD define rules for handling
+  token will be encrypted using a key known only to the server, thereby
+  conforming to the requirements in {{validate-future}}. After that, the server
+  sends a NEW_TOKEN frame that contains the generated token and the seeds that
+  have been embedded to that token.
```
```suggestion
  have been embedded into that token.
```

mikkelfj commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2724,96 @@ between endpoints.  Application protocols SHOULD define rules for handling
+  version number, the server embeds the NEW_TOKEN token found in the Initial
+  packet within the retry token it issues.  Once the server receives a response
+  from the client carrying that retry token and the path is validated, it
+  decrypts the NEW_TOKEN token embedded in the retry token to recover the
```
```suggestion
  decrypts the NEW_TOKEN token embedded within the retry token to recover the
```

mikkelfj commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2724,96 @@ between endpoints.  Application protocols SHOULD define rules for handling
+list to decrypt the payload of Initial packets using those alternative version
+numbers.  But still, having a set of version numbers and initial salts used
+concurrently is considered better than just using the default values of QUIC in
+terms of preventing ossification.
```
I am not comfortable with the slack given here. While it does improve things when salt is reused, it also gives the client a false sense of security. It would be better if the client could be certain that it does indeed sent private information on the wire.

For example it might create an initial connection to bootstrap safety, similar to if it connected to a secure DNS with crypto keys (TBD), and then reconnects to send confidential informational information. 

Taking this one further, it would be advantageous if a token could be used across different server certificates so the initial certificate could be "harmless" in order to enhance privacy.

mikkelfj commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2724,96 @@ between endpoints.  Application protocols SHOULD define rules for handling
+
+* Version number; a 32-bit unsigned number that is to be presented on wire in
+  place of the version number specified in this document.  This value MUST NOT
+  be a reserved version ({{versions}}).
```
It is not very clear from this text if the intention is to generate a new representation of the same version in which the token was generated, or there is indeed a new version. Since this does not cover version negotiation, it presumably refers to the same version. But that could change as versions get added and perhaps V1 should be prepared for that (I'm not sure I am getting this right). @MikeBishop alluded to something similar in an earlier comment AFAIR.

martinthomson replied:
I don't think that it is necessary to prohibit the use of any allocated version, including the reserved versions.

kazuho commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2724,96 @@ between endpoints.  Application protocols SHOULD define rules for handling
+
+This specification defines how the seeds are used as well as how they are
+advertised using a NEW_TOKEN frame.  Other specifications MAY define other
+methods for distributing or deducing these seeds.
```
HTTP/3 defining Alt-Svc, SVCB / HTTPSSVC DNS records, ESNI, to name a few.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2724,96 @@ between endpoints.  Application protocols SHOULD define rules for handling
+Typically, a server that advertises the alternative seeds would act in the
+following steps:
+
+* The server pre-allocates a set of unused version numbers as the alternative
```
This implies that you can't provide a different salt for version 1.  But I am fairly certain that you can with this design.  The same applies to ?a?a?a?a.

If you only support versions that include tokens, then those tokens can encode the real version and you can fill the real version number with noise (that you should still authenticate, of course).

If there are different real versions that carry tokens in different places, you might have some trouble finding tokens.  But I believe that this is achievable if you reorder or restructure token contents.

If a real version you support doesn't carry a token in the Initial-equivalent packet, then you just need to avoid clobbering that version (or versions, if it uses a different aliasing scheme).

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2724,96 @@ between endpoints.  Application protocols SHOULD define rules for handling
+## Server Behavior
+
+Typically, a server that advertises the alternative seeds would act in the
+following steps:
```
I think that this advice is too constraining.  I would instead say, probably as the sum of total of this section:


> To protect against the possibility of third parties fixating on specific aspects of the encoding of unprotected Initial packets, an optional version aliasing function is provided that is bound to address validation tokens provided using NEW_TOKEN.
>
> A server that supports version aliasing includes an alias version, a packet type mask, and replacement Initial salt with every NEW_TOKEN message.  This provides obfuscation of Initial packets for subsequent connections to the same server. Use of this capability is optional for servers.  A server MAY use default values to avoid creating aliased versions.
>
> A client MUST use aliased versions if tokens from NEW_TOKEN are used.  Details of how version alias fields are used are included in {{frame-new-token}}.
>
> A server commits to supporting an aliased version for the entire time that the token is valid (as indicated by the Lifetime field). Therefore, servers need to ensure that the process used to select version alias fields is the same for all servers that might accept the token, both across different server instances in the same deployment and across potential crashes or reboots that might cause servers to lose state.
>
> The purpose of version aliases are to confound attempts by third parties to fixate on a specific form of Initial packets. Therefore, the information that a server uses to interpret aliased versions does not need to be based on secret information and it does not have to change frequently.  Consequently, building values into software is a viable means of ensuring that values are not lost.
>
> As an example of an implementation that is both robust against loss of state and still difficult for third parties to fixate on, a server could generate an aliased version and associated values using a pseudorandom function that uses a static key.  Including a key identifier in the token should allow the server to identify the key that is in use, which allows for periodic key rotation.  As long as old keys are retained until all tokens issued using that key have expired, the server is able to correctly recognize any valid aliased version.
>
> A server that provides aliased versions needs to remove any version aliasing prior to determining whether to send a Version Negotiation packet when processing incoming packets as described in {{server-pkt-handling}}.

I think that you might also want to update the text in [Section 5.2.2](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#server-pkt-handling) to refer to this new section.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2724,96 @@ between endpoints.  Application protocols SHOULD define rules for handling
+
+* When sending a Retry in response to an Initial packet carrying an alternative
+  version number, the server embeds the NEW_TOKEN token found in the Initial
+  packet within the retry token it issues.  Once the server receives a response
```
This is an important design point that is hidden: the aliased version applies to all Initial packets that are sent.  That means that the alternative salt is used when the server provides an updated connection ID.  Do the Handshake and 0-RTT packets also use the aliased version?  That needs to be right up front.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2724,96 @@ between endpoints.  Application protocols SHOULD define rules for handling
+
+This specification defines how the seeds are used as well as how they are
+advertised using a NEW_TOKEN frame.  Other specifications MAY define other
+methods for distributing or deducing these seeds.
```
I don't think that we need to say this here.  I think that this subsection can be removed.

kazuho commented on draft-ietf-quic-transport.md:
```diff
@@ -2721,6 +2724,96 @@ between endpoints.  Application protocols SHOULD define rules for handling
+A server (or a set of servers) that cannot satisfy this requirement can stick to
+using the default values by consistently advertising the default version number,
+default initial salt, and a packet type modifier of zero in the NEW_TOKEN frames
+it sends.
```
Am I correct in assuming that what you are looking for is a clarification that servers can announce a set of alternative versions, each of them tied to a fixed initial salt and packet type modifier?

I think that this is the correct text explaining the multi-CDN use-case, where servers cannot agree on anything.

To paraphrase, we'd be explaining:
1. a server that generates a new initial salt for every connection and embeds it into NEW_TOKEN
2. a server that has a fixed mapping between alternative versions vs. salt (and packet type modifier).
3. a server that does not do greasing at all (multi-CDN).

martinthomson commented:
With the closing of the related issues, this can be closed too.
